File: moling-server/app/dao/notification_dao.py
"""墨灵 (Moling) — Notification DAO."""

from __future__ import annotations

from sqlalchemy import select, func, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.dao.base_dao import BaseDAO
from app.models.notification import Notification


class NotificationDAO(BaseDAO[Notification]):
    """Data access object for Notification model."""

    def __init__(self):
        super().__init__(Notification)

    async def get_by_user(
        self,
        db: AsyncSession,
        user_id: int,
        *,
        skip: int = 0,
        limit: int = 20,
        is_read: bool | None = None,
    ) -> list[Notification]:
        """Get notifications for a specific user with pagination."""
        stmt = select(Notification).where(Notification.user_id == user_id)
        
        if is_read is not None:
            stmt = stmt.where(Notification.is_read == is_read)
        
        stmt = stmt.order_by(Notification.created_at.desc())
        stmt = stmt.offset(skip).limit(limit)
        
        result = await db.execute(stmt)
        return list(result.scalars().all())

    async def count_by_user(
        self,
        db: AsyncSession,
        user_id: int,
        *,
        is_read: bool | None = None,
    ) -> int:
        """Count notifications for a specific user."""
        stmt = select(func.count()).select_from(Notification).where(
            Notification.user_id == user_id
        )
        
        if is_read is not None:
            stmt = stmt.where(Notification.is_read == is_read)
        
        result = await db.execute(stmt)
        return result.scalar_one()

    async def mark_as_read(
        self,
        db: AsyncSession,
        notification_id: int,
        user_id: int,
    ) -> Notification | None:
        """Mark a notification as read (with user ownership check)."""
        stmt = (
            update(Notification)
            .where(Notification.id == notification_id, Notification.user_id == user_id)
            .values(is_read=True)
            .returning(Notification)
        )
        result = await db.execute(stmt)
        return result.scalar_one_or_none()

    async def mark_all_as_read(
        self,
        db: AsyncSession,
        user_id: int,
    ) -> int:
        """Mark all notifications as read for a user. Returns count of updated."""
        stmt = (
            update(Notification)
            .where(Notification.user_id == user_id, Notification.is_read == False)
            .values(is_read=True)
        )
        result = await db.execute(stmt)
        return result.rowcount

    async def delete_by_user(
        self,
        db: AsyncSession,
        notification_id: int,
        user_id: int,
    ) -> Notification | None:
        """Delete a notification (with user ownership check)."""
        stmt = select(Notification).where(
            Notification.id == notification_id,
            Notification.user_id == user_id,
        )
        result = await db.execute(stmt)
        notification = result.scalar_one_or_none()
        
        if notification:
            await db.delete(notification)
            await db.flush()
        
        return notification


# Singleton instance
notification_dao = NotificationDAO()

File: moling-server-rs/crates/moling-db/src/dao/notification_dao.rs
//! Notification DAO — notification CRUD with read-status management.

use moling_core::error::{AppError, AppResult};
use sea_orm::{ActiveModelTrait, ColumnTrait, DatabaseConnection, EntityTrait, PaginatorTrait, QueryFilter, QueryOrder, QuerySelect};

use crate::entities::notification::{self, ActiveModel, Entity as Notification, Model as NotificationModel};

/// Notification data access object.
#[derive(Clone, Default)]
pub struct NotificationDao;

impl NotificationDao {
    /// Find a notification by ID.
    pub async fn find_by_id(
        &self,
        db: &DatabaseConnection,
        id: i32,
    ) -> AppResult<Option<NotificationModel>> {
        Notification::find_by_id(id)
            .filter(notification::Column::IsDeleted.eq(false))
            .one(db)
            .await
            .map_err(|e| {
                tracing::error!(%id, "Notification: database error finding by id: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// List notifications for a user with pagination and optional read filter.
    pub async fn list_by_user(
        &self,
        db: &DatabaseConnection,
        user_id: &str,
        skip: u64,
        limit: u64,
        is_read: Option<bool>,
    ) -> AppResult<Vec<NotificationModel>> {
        let mut query = Notification::find()
            .filter(notification::Column::UserId.eq(user_id))
            .filter(notification::Column::IsDeleted.eq(false));

        if let Some(read) = is_read {
            query = query.filter(notification::Column::IsRead.eq(read));
        }

        query
            .order_by_desc(notification::Column::CreatedAt)
            .offset(skip)
            .limit(limit)
            .all(db)
            .await
            .map_err(|e| {
                tracing::error!(user_id, "Notification: database error listing by user: {e}");
                AppError::internal("Database query failed")
            })
    }

    /// Count notifications for a user, with optional read filter.
    pub async fn count_by_user(
        &self,
        db: &DatabaseConnection,
        user_id: &str,
        is_read: Option<bool>,
    ) -> AppResult<u64> {
        let mut query = Notification::find()
            .filter(notification::Column::UserId.eq(user_id))
            .filter(notification::Column::IsDeleted.eq(false));

        if let Some(read) = is_read {
            query = query.filter(notification::Column::IsRead.eq(read));
        }

        query.count(db).await.map_err(|e| {
            tracing::error!(user_id, "Notification: database error counting: {e}");
            AppError::internal("Database query failed")
        })
    }

    /// Create a new notification.
    pub async fn create(
        &self,
        db: &DatabaseConnection,
        model: ActiveModel,
    ) -> AppResult<NotificationModel> {
        model.insert(db).await.map_err(|e| {
            tracing::error!("Notification: database error creating: {e}");
            AppError::internal("Database insert failed")
        })
    }

    /// Mark a single notification as read (with user ownership check).
    pub async fn mark_as_read(
        &self,
        db: &DatabaseConnection,
        notification_id: i32,
        user_id: &str,
    ) -> AppResult<Option<NotificationModel>> {
        use sea_orm::{IntoActiveModel, Set};
        let entity = Notification::find()
            .filter(notification::Column::Id.eq(notification_id))
            .filter(notification::Column::UserId.eq(user_id))
            .one(db)
            .await
            .map_err(|e| {
                tracing::error!(%notification_id, "Notification: database error finding for mark-read: {e}");
                AppError::internal("Database query failed")
            })?
            .ok_or_else(|| AppError::not_found("通知不存在"))?;

        let mut active = entity.into_active_model();
        active.is_read = Set(true);
        let updated = active.update(db).await.map_err(|e| {
            tracing::error!(%notification_id, "Notification: database error marking as read: {e}");
            AppError::internal("Database update failed")
        })?;
        Ok(Some(updated))
    }

    /// Mark all unread notifications as read for a user.
    /// Returns the number of notifications updated.
    pub async fn mark_all_as_read(
        &self,
        db: &DatabaseConnection,
        user_id: &str,
    ) -> AppResult<u64> {
        use sea_orm::sea_query::Expr;
        let result = Notification::update_many()
            .filter(notification::Column::UserId.eq(user_id))
            .filter(notification::Column::IsRead.eq(false))
            .col_expr(notification::Column::IsRead, Expr::value(true))
            .exec(db)
            .await
            .map_err(|e| {
                tracing::error!(user_id, "Notification: database error marking all as read: {e}");
                AppError::internal("Database update failed")
            })?;
        Ok(result.rows_affected)
    }

    /// Hard-delete a notification (with user ownership check).
    pub async fn delete_by_user(
        &self,
        db: &DatabaseConnection,
        notification_id: i32,
        user_id: &str,
    ) -> AppResult<Option<NotificationModel>> {
        use sea_orm::ModelTrait;
        let entity = Notification::find()
            .filter(notification::Column::Id.eq(notification_id))
            .filter(notification::Column::UserId.eq(user_id))
            .one(db)
            .await
            .map_err(|e| {
                tracing::error!(%notification_id, "Notification: database error finding for delete: {e}");
                AppError::internal("Database query failed")
            })?;

        if let Some(e) = entity {
            let deleted = e.clone();
            e.delete(db).await.map_err(|err| {
                tracing::error!(%notification_id, "Notification: database error deleting: {err}");
                AppError::internal("Database delete failed")
            })?;
            Ok(Some(deleted))
        } else {
            Ok(None)
        }
    }
}
